Do not try to infer the logprob of broadcasted or advanced indexed variables

ricardoV94 opened the pull request:
This PR removes an invalid rewrite from the logprob submodule

It also suppresses a NotImplementedError in the Mixture rewrite. Individual rewrites should not judge whether logprob for a graph is implemented or not, because other rewrites might be able to handle it.

ricardoV94 committed:
Update xfailed invalid broadcast test

The test now failed due to improvement in static type shapes, but the underlying issue was not addressed. The logp for a broadcasted RV must consider that broadcasted dimensions are not independent

ricardoV94 committed:
Do not raise from internal mixture rewrite

ricardoV94 commented on pymc/tests/logprob/test_transforms.py:
```diff
@@ -709,13 +709,13 @@ def test_discrete_rv_multinary_transform_fails():
 
 
-@pytest.mark.xfail(reason="Check not implemented yet, see #51")
+@pytest.mark.xfail(reason="Check not implemented yet")
```
Add companion test with explicit `broadcast_to` which now fails after this PR

ricardoV94 commented on pymc/logprob/mixture.py:
```diff
@@ -407,6 +418,8 @@ def logprob_MixtureRV(
 
+    # FIXME: This logic does not account for repeated indexes, in which case the Mixture values are not independent.
+    #  It over-counts the logp terms for values that were repeated during indexing
     if len(indices) > 1 or indices[0].ndim > 0:
```
Make sure we have tests for still supported multiple indexes

```
Normal.dist(size=(2, 3)[0, 0]
```

ricardoV94 replied:
Remove or branch (Advanced Indexing)?

zaxtax commented on pymc/logprob/mixture.py:
```diff
@@ -300,13 +300,27 @@ def mixture_replace(fgraph, node):
-    mixture_res, join_axis = get_stack_mixture_vars(node)
+    try:
+        mixture_res, join_axis = get_stack_mixture_vars(node)
+    except NotImplementedError:
```
This seems like a code smell. Would it be unreasonable to refactor `get_stack_mixture_vars` to not raise this exception?

ricardoV94 replied:
I refactored it to avoid raising the exception

zaxtax commented on pymc/logprob/mixture.py:
```diff
@@ -404,6 +418,8 @@ def logprob_MixtureRV(
     assert len(indices) > 0
 
+    # FIXME: This logic does not account for repeated indexes, in which case the Mixture values are not independent.
+    #  It over-counts the logp terms for values that were repeated during indexing
```
Should this have a unit test when it's resolved?

ricardoV94 replied:
I don't know if we will ever resolve it. For now we don't allow creating such MixtureRVs in the first place. The ones that hit this branch don't do AdvancedIndexing

ricardoV94 committed:
Reorganize and label Mixture test parametrizations

ricardoV94 committed:
Do not treat broadcasted variables as independent in logprob inference

This included two cases:

## Direct valuation of broadcasted RVs

The `naive_bcast_lift` rewrite was included by default and allowed broadcasted RVs to be valued. This is invalid because it implies that `logp(broadcast_to(normal(0, 1), (3, 2), value) == logp(normal(0, 1, size=(3, 2)), value)` which is not true. Broadcast replicates the same RV draws, so these values can't be considered independent when evaluating the logp.

The rewrite is kept but not used anywhere

## Valuation of Mixtures with potential repeated components

This can happen when AdvancedIndexing is used. As a precaution, Mixture replace now fails when Advanced integer indexing is detected, even though some cases may be valid at runtime (e.g., no repated indexes)

ricardoV94 committed:
Add failing test when IfElse Mixture logprob is used with indexing that mixes across components
